Dev flow.to_global()/.to_local() to Support Multiple Input Types

lixiang007666 opened the pull request:
### This PR is done:

Refine @Alive1024 的 flow.to_global/.to_local 增加多种输入类型的输入支持（包括 state dict）的 PR，增加 dict_to_global op，支持按 s 分片来 save 和 load 大模型，并推进合并。

- [x] Fork branch: https://github.com/Oneflow-Inc/oneflow/pull/8091
- [x] Opt dict_to_global
- [x] Resolve conflict
- [x] Refine
- [x] Add Docs

**8.23 更新：删除 dict_to_global 接口，改用 https://github.com/Oneflow-Inc/oneflow/pull/8868#issuecomment-1222012217 里提到的方式。**


### Doc Preview:

增加 utils.global_view 和 utils.tensor，作用域风格和 torch 一致。

#### flow.utils.global_view.to_global:
<img width="678" alt="image" src="https://user-images.githubusercontent.com/54010254/186127674-6969fd4a-8350-4952-af84-070f759667a8.png">
<img width="625" alt="image" src="https://user-images.githubusercontent.com/54010254/186127749-f453611a-63e8-4e99-8c28-3961dbfbca23.png">
<img width="612" alt="image" src="https://user-images.githubusercontent.com/54010254/186127783-dc708483-c8db-4821-ae31-2df21806c71d.png">


#### flow.utils.global_view.to_local:
<img width="577" alt="image" src="https://user-images.githubusercontent.com/54010254/185062031-6edff1f9-f667-440a-92a4-2ac749f2c42b.png">

lixiang007666 commented on python/oneflow/nn/modules/global_cast.py:
```diff
@@ -74,17 +81,358 @@ def global_to_global_op(
+    return flow._C.to_local(input_tensor, copy)
 
 
 def to_local_op(input, *, copy=False):
```
这个接口增加了 copy 参数。

lixiang007666 commented on python/oneflow/nn/modules/global_cast.py:
```diff
@@ -74,17 +81,358 @@ def global_to_global_op(
+    return output_dict
+
+
+def dict_to_global(input_dict, placement, sbp, *, sbp_for_special_keys):
```
https://github.com/Oneflow-Inc/oneflow/pull/8091#issuecomment-1198909542 ，这个 comment 提到的问题已经被解决， dict_to_global 更新为和 to_global 接口一样的实现思路，只是在调用 _to_global_tensor  时，增加了 sbp=sbp_for_special_keys[key] 的处理。

lixiang007666 commented on python/oneflow/test/graph/test_graph_save_load_global_b_s.py:
```diff
@@ -0,0 +1,808 @@
+            )
+            if flow.env.get_rank() in model_file_placement.ranks:
+                flow.save(flow.to_local(model_file_state_dict), state_dict_dir)
+
```
在两个设备上做流水并行，model_file_placement 为 oneflow.placement("cpu", ranks=[0, 1])。
在这里保存模型。

lixiang007666 commented on python/oneflow/test/graph/test_graph_save_load_global_b_s.py:
```diff
@@ -0,0 +1,808 @@
+                sbp=flow.sbp.split(0),
+                sbp_for_special_keys=sbp_for_special_keys,
+            )
+            graph_model.load_state_dict(global_state_dict)
```
在两个设备上做流水并行，model_file_placement 为 oneflow.placement("cpu", ranks=[0, 1])。
在这里加载模型。

lixiang007666 commented on python/oneflow/test/graph/test_graph_save_load_global_b_s.py:
```diff
@@ -0,0 +1,808 @@
+            )
+            if flow.env.get_rank() in model_file_placement.ranks:
+                flow.save(flow.to_local(model_file_state_dict), state_dict_dir)
+
```
在四个设备上做流水并行，model_file_placement 为 oneflow.placement("cpu", ranks=[0, 1])。
在这里保存模型。

lixiang007666 commented on python/oneflow/test/graph/test_graph_save_load_global_b_s.py:
```diff
@@ -0,0 +1,808 @@
+                sbp_for_special_keys=sbp_for_special_keys,
+            )
+            graph_model.load_state_dict(global_state_dict)
+
```
在四个设备上做流水并行，model_file_placement 为 oneflow.placement("cpu", ranks=[0, 1])。
在这里加载模型。

lixiang007666 commented on python/oneflow/test/graph/test_graph_save_load_global_b_s.py:
```diff
@@ -0,0 +1,808 @@
+                local_state_dict, placement=model_file_placement, sbp=B
+            )
+            linear_t_g.load_state_dict(global_state_dict)
+
```
在两个设备上做数据并行，model_file_placement 为 oneflow.placement("cpu", ranks=[0])。
在这里加载模型。
使用 to_global ，后面的 `test_graph_save_load_global_split`都是 dict_to_global。

lixiang007666 commented on python/oneflow/test/graph/test_graph_save_load_global_b_s.py:
```diff
@@ -0,0 +1,808 @@
+            if cur_rank in model_file_placement.ranks:
+                iter1_local_dict = flow.to_local(model_file_state_dict)
+                flow.save(iter1_local_dict, state_dict_dir)
+
```
在两个设备上做数据并行，model_file_placement 为 oneflow.placement("cpu", ranks=[0])。
在这里保存模型。

lixiang007666 commented on python/oneflow/test/modules/test_global_cast.py:
```diff
@@ -1286,5 +1288,206 @@ def test_global_to_global_1to1_cpu_to_gpu(test_case):
+                if isinstance(node, Tensor):
+                    test_case.assertFalse(node.is_global)
+
+    def _test_any_input(test_case):
```
to_global 支持 tensor list tuple 和 dict。

lixiang007666 commented on python/oneflow/test/modules/test_global_cast.py:
```diff
@@ -1286,5 +1288,206 @@ def test_global_to_global_1to1_cpu_to_gpu(test_case):
+            ret = flow.to_local(i)
+            test_case.__all_local(ret)
+
+    def _test_tensor_to_global(test_case):
```
测试 to_global 输入为 tensor 的情况，分四种：local tensor -> global tensor；global tensor -> global tensor；global tensor -> global tensor，但是 P 和 SBP 为空；global tensor -> global tensor，但是 SBP 错误，这时应该抛出 TypeError。

lixiang007666 commented on python/oneflow/test/modules/test_global_cast.py:
```diff
@@ -1286,5 +1288,206 @@ def test_global_to_global_1to1_cpu_to_gpu(test_case):
+                sbp=(TestToGlobalAndLocal.sbp, 0),
+            )
+
+    def _test_tensor_to_local(test_case):
```
测试 global tensor -> local tensor。

lixiang007666 commented on python/oneflow/test/modules/test_global_cast.py:
```diff
@@ -1286,5 +1288,206 @@ def test_global_to_global_1to1_cpu_to_gpu(test_case):
+        # local input, display warning
+        local_state_dict = flow.to_local(local_state_dict)
+
+    def _test_eagar_state_dict(test_case):
```
在 eager 下，测试 to_global 和 to_local 的参数类型为 state_dict 的情况，

lixiang007666 commented on python/oneflow/test/modules/test_global_cast.py:
```diff
@@ -1286,5 +1288,206 @@ def test_global_to_global_1to1_cpu_to_gpu(test_case):
+        )
+        test_case.__test_state_dict_to_local(
+            TestToGlobalAndLocal.global_graph_model.state_dict()
+        )
```
local_graph_model = GraphTestModel(model)
global_graph_model = GraphTestModel(model.to_global(...))

lixiang007666 commented on python/oneflow/nn/modules/global_cast.py:
```diff
@@ -74,17 +81,358 @@ def global_to_global_op(
+        placement (oneflow.placement): the common placement for all keys.
+        sbp (oneflow.sbp.sbp): the default SBP for not special keys.
+        sbp_for_special_keys (dict): The keys are str type, and the values are oneflow.sbp.sbp type.
+         This is used to specify special SBPs for "System-Train-TrainStep" tensor with shape [1] etc.
```
Ex: `flow.dict_to_global(state_dict, placement=model_file_placement, sbp=flow.sbp.split(0),sbp_for_special_keys={"System-Train-TrainStep": flow.sbp.broadcast})`

strint replied:
dict_to_global 这个可以先放到 oneflow.utils.global 下面，作为辅助工具

strint commented on python/oneflow/utils/to_global/dict_to_global.py:
```diff
@@ -0,0 +1,139 @@
+"""
```
改为 utils.global 吧，放一些和 global 关联的复杂工具

lixiang007666 replied:
这个我改成 utils.global_utils 了，global 是 py 关键字不能用好像～

strint replied:
global_view 吧

strint commented on python/oneflow/nn/modules/global_cast.py:
```diff
@@ -74,17 +81,245 @@ def global_to_global_op(
+        Both placement and sbp are required if the input is local, otherwise at least one of placement and sbp is required.
+
+    Args:
+        input (oneflow.Tensor/None/list/tuple/dict): the input that needs to be converted.
```
和 list/tuple/dict 的 to_global/to_local 关联的功能，都放到 utils.global 下面吧

strint replied:
global_cast 下面的原接口还是保持简单

strint replied:
to_global_op 不做改动，把这里的改动，都放到 global_view 下面把

lixiang007666 commented on docs/source/oneflow.rst:
```diff
@@ -386,3 +386,14 @@ BLAS and LAPACK Operations
+
+    to_global
+    to_local
+    dict_to_global
```
这个文档的位置还没修改。

strint commented on docs/source/utils.rst:
```diff
@@ -0,0 +1,16 @@
+    :toctree: generated
+    :nosignatures:
+
+    from_torch
```
oneflow.utils.tensor.[from_torch, to_torch]

strint commented on docs/source/utils.rst:
```diff
@@ -0,0 +1,16 @@
+
+    from_torch
+    to_torch
+    to_global
```
oneflow.utils.global_view.[...]

strint commented on python/oneflow/utils/global_view/to_global.py:
```diff
@@ -0,0 +1,144 @@
+    Returns:
+        The converted input.
+
+    For a tensor input: please refer to the examples in :func:`oneflow.Tensor.to_global`.
```
补充一个 Note：
如果 dict 类型，比如 state dict， to_global 时无法使用统一的 sbp 时，推荐使用 dict_to_global.

strint commented on python/oneflow/utils/global_view/dict_to_global.py:
```diff
@@ -0,0 +1,139 @@
+        placement (oneflow.placement): the common placement for all keys.
+        sbp (oneflow.sbp.sbp): the default SBP for not special keys.
+        sbp_for_special_keys (dict): The keys are str type, and the values are oneflow.sbp.sbp type.
+         This is used to specify special SBPs for "System-Train-TrainStep" tensor with shape [1] etc.
```
想整体 to split(0) ，但是存在 dim 0 无法切分的 tensor，那么这些 tensor 可以指定专用的 sbp

strint commented on python/oneflow/utils/global_view/dict_to_global.py:
```diff
@@ -0,0 +1,139 @@
+        sbp_for_special_keys (dict): The keys are str type, and the values are oneflow.sbp.sbp type.
+         This is used to specify special SBPs for "System-Train-TrainStep" tensor with shape [1] etc.
+         Also, it is worth noting that, for a tensor of shape `(1, n)`, you can specify SBP is `oneflow.sbp.split(1)`.
+         The keys here support the use of "." to indicate nested dicts.
```
这里看起来，graph 的 state dict，最好还是打平 key，而不使用嵌套，比较一致。

strint commented on python/oneflow/test/graph/test_graph_save_load_global_b_s.py:
```diff
@@ -0,0 +1,808 @@
+                placement=model_file_placement,
+                sbp=flow.sbp.split(0),
+                sbp_for_special_keys=sbp_for_special_keys,
+            )
```
分片 Load step 2、在部分 rank 上按 split to_global

strint commented on python/oneflow/test/graph/test_graph_save_load_global_b_s.py:
```diff
@@ -0,0 +1,808 @@
+            if cur_rank in model_file_placement.ranks:
+                local_state_dict = flow.load(state_dict_dir)
+            else:
+                local_state_dict = None
```
分片 Load step 1、在部分 rank 上按 split load 模型

strint commented on python/oneflow/test/graph/test_graph_save_load_global_b_s.py:
```diff
@@ -0,0 +1,808 @@
+        iter1_state_dict = graph_model.state_dict()
+
+        if call_cnt == 0:
+            model_file_state_dict = flow.utils.dict_to_global(
```
分片 Save step 1、按 split to_global 到部分 rank

strint commented on python/oneflow/test/graph/test_graph_save_load_global_b_s.py:
```diff
@@ -0,0 +1,808 @@
+                sbp_for_special_keys=sbp_for_special_keys,
+            )
+            if flow.env.get_rank() in model_file_placement.ranks:
+                flow.save(flow.utils.to_local(model_file_state_dict), state_dict_dir)
```
分片 Save step 2，按 local tensor 保存到部分 rank 上；

strint commented:
这PR支持了tupe/list/dict 的 to_global/to_local，作为 global view 辅助方法（utils)。

现有的 flow.save/load 的 global 模式，默认是在一个 rank 上 按 Broadcast 来保存加载模型。

这里 dict_to_global 提高了 save/load 的灵活性，可以支持在任意 placement + 任意 sbp。其中一个主要应用场景是支持大模型的分片保存。可以在几个 rank 上，按 Split 来保存和加载模型。

lixiang007666 commented:
## DeepSpeed 分片保存和加载模型初步调研

DeepSpeed 训练出的 Turing-NLG 有 17B 参数量，DeepSpeed 和 Megatron 驱动 MT-NLG 具有 530B 个参数。支持这样的大规模语言模型需要分片保存和加载，不会使用单机内存。

我们目前采用的方式是支持按 s 分片来保存和加载模型，举个例子：

```python
# 将 state dict 按 s 分片到 model_file_placement 上。
model_file_state_dict = flow.utils.global_view.to_global(
    iter1_state_dict, placement=model_file_placement, sbp=flow.sbp.split(0)
    )
# 保存模型：
if flow.env.get_rank() in model_file_placement.ranks:
    flow.save(flow.utils.global_view.to_local(model_file_state_dict), state_dict_dir)
# 加载模型：
if cur_rank in model_file_placement.ranks:
    local_state_dict = flow.load(state_dict_dir)
else:
    local_state_dict = None
global_state_dict = flow.utils.global_view.dict_to_global(
    local_state_dict,
    placement=model_file_placement,
    sbp=flow.sbp.split(0),
)
# 对 global_state_dict 后续操作
```

DeepSpeed 目前支持超大模型的分片存储，既分片检查点格式，

下面简单看Megatron-DeepSpeed 仓库里 TP（TensorParallel） 下是如何将分片文件（Tensor）合并的。 

相关函数：[https://github.com/bigscience-workshop/Megatron-DeepSpeed/blob/0f23a729ec5ed6ef46e1fecc8071ac11acdb91c6/tools/convert_checkpoint/ds_to_universal.py#L221](https://github.com/bigscience-workshop/Megatron-DeepSpeed/blob/0f23a729ec5ed6ef46e1fecc8071ac11acdb91c6/tools/convert_checkpoint/ds_to_universal.py#L221)

```python
def _merge_tp_slice_files(args, ds_checkpoint, slice_shapes, temp_dir):
    work_chunks = list(_get_chunks(list(slice_shapes.items()), args.num_merge_workers))
    #pprint(work_chunks)
    zero_output_folder = os.path.join(args.output_folder, "zero") 
    do_work = partial(merge_tp_slices, ds_checkpoint, zero_output_folder, temp_dir, ds_checkpoint.tp_degree)
    _do_parallel_work(do_work, work_chunks, args.num_merge_workers)

def merge_tp_slices(ds_checkpoint, dir, slice_dir, tp_degree, name_and_shape):
    name, shape = name_and_shape
    slice_base_path = os.path.join(slice_dir, name)
    param_base_path = os.path.join(dir, name)

    for state in ("fp32", "exp_avg", "exp_avg_sq"):
        slices = _merge_zero_shards(slice_base_path, state, tp_degree, shape)
        final_path = os.path.join(param_base_path, f"{state}.pt")

        #print(f"Expected shape: {shape}")
        #print(f"Fragment sizes:", list(frag.shape for frag in slices))
        ckpt_dict = {}
        if any(re.match(pattern, name) for pattern in WEIGHTS_TO_AVERAGE_PATTERNS):
            param = sum(slices) / len(slices)
        else:
            cat_dim = 1 if any(text in name for text in WEIGHTS_WITH_ROW_PARALLELISM_CONTAIN) else 0
            #print(f"CAT DIM: {cat_dim}")
            param = torch.cat(slices, dim=cat_dim)
            ckpt_dict['cat_dim'] = cat_dim

        if "word_embeddings.weight" in name:
            #print(f"Before {param.shape=}")
            # strip padding
            #param = _strip_vocab_padding(ds_checkpoint, param)
            ckpt_dict['vocab_divisibility_padding_tensor'] = _get_vocab_divisibility_padding_tensor(ds_checkpoint, param)
            #print(f"After {param.shape=}")

        #print(f"Final shape: {param.shape}")
        ckpt_dict['param'] = param
        _save_checkpoint(final_path, ckpt_dict)
```

先看 _merge_tp_slice_files，只需要关注该函数将以多线程的方式调用 merge_tp_slices，并且 work_chunks 传入 merge_tp_slices 的 name_and_shape 参数。在 merge_tp_slices 中，ds_checkpoint 为 DeepSpeedCheckpoint 类对象（ [https://github.com/microsoft/DeepSpeed/blob/fda63432ba67643d2f54885e45f0b2058fa2b937/deepspeed/checkpoint/deepspeed_checkpoint.py](https://github.com/microsoft/DeepSpeed/blob/fda63432ba67643d2f54885e45f0b2058fa2b937/deepspeed/checkpoint/deepspeed_checkpoint.py#L37) ），ds_checkpoint.tp_degree 为并行设备数量。主要逻辑是，首先遍历三种 "fp32", "exp_avg", "exp_avg_sq" 权重，每种权重又分别保存检查点。_merge_zero_shards 具体用法是根据 slice_base_path + tp_index（并行设备 id）+ 传入的三种 state 组成的路径，通过分片路径返回所有切片，slices 这时是一个经过 reshape 的简单 1d 数组。

再往下看，可以发现初始化的 final_path 已经没有了 tp_index，param_base_path 是我们指定的 zero_output_folder。在后面的逻辑中，处理 param 的几种情况，包括将上面得到的 slices 做拼接。

再关注 [https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/checkpoint/zero_checkpoint.py](https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/checkpoint/zero_checkpoint.py) 和 [https://github.com/microsoft/DeepSpeed/blob/fda63432ba67643d2f54885e45f0b2058fa2b937/deepspeed/checkpoint/deepspeed_checkpoint.py](https://github.com/microsoft/DeepSpeed/blob/fda63432ba67643d2f54885e45f0b2058fa2b937/deepspeed/checkpoint/deepspeed_checkpoint.py#L72) 两个文件，DeepSpeed 是支持 TP+PP+DP 来完成分片检查点，其中 PP 检查点和和 ZeRO 检查点略有不同，比如它将每一层都放在一个单独的检查点中，并将每个 TP 拆分保存在一个单独的检查点中。这里暂时我是没有关注的，后面有时间可以再看。

#### Conclusion：
也就是说，目前 OneFlow 做分片保存，也是需要考虑路径的（比如上面的 tp_index 用法），一个 rank 对应一个路径，路径名用 rank id 的方式。

strint commented:
TODO：
- [ ] 1、修改下测试的名字，2n2s 这样的名字太简单了，写的具体一点；
- [ ] 2、多 rank 分片保存之前使用同一个目录，没有出问题的原因，注释一下，然后提供正确路径生成方法；

lixiang007666 commented on python/oneflow/test/graph/test_graph_save_load_global_b_s.py:
```diff
@@ -0,0 +1,817 @@
+    rank_id = flow.env.get_rank()
+    with tempfile.TemporaryDirectory(
+        prefix="graph_save_load_global_" + str(rank_id)
+    ) as state_dict_dir:
```
这里之前没有加上 rand_id（多 rank 分片保存看似是一个目录），之所以没有报错是因为 tempfile 每次会生成不同的文件名后缀，所以没有覆盖。
正确做法是现在的方法，提供给使用的使用方法在：https://github.com/Oneflow-Inc/oneflow/pull/8868#issuecomment-1221671880

lixiang007666 commented on python/oneflow/test/graph/test_graph_save_load_global_b_s.py:
```diff
@@ -0,0 +1,817 @@
+@unittest.skipIf(os.getenv("ONEFLOW_TEST_CPU_ONLY"), "only test cpu cases")
+@flow.unittest.skip_unless_1n4d()
+class TestGraphSaveLoadGlobal4d(oneflow.unittest.TestCase):
+    def test_graph_save_load_gpu_2_split_2_none(test_case):
```
这个测试函数名改了，表示 4 个rank下，在前面两个 rank 为 s，后面两个为 none。前面的文件名也对应纠正过了。

daquexian commented:
分 dict_to_global 和 global 两个接口是不是不太好，并且不少用户把 optimizer state 和 model state 套在一个大 dict 里保存加载，就没办法用 state_dict 的 key 来直接定位到 tensor 了

如果把 `flow.utils.dict_to_global(x)` 合并进 `flow.utils.to_global(x)`，并支持用户传进来一个 `(x, tensor) -> sbp` 的函数来解决不同 tensor 不同 sbp 的需求呢：

```python
state_dict = flow.load(...)

def get_sbp(state_dict, tensor):
    if tensor is state_dict["System-Train-TrainStep"]:
        return BROADCAST
    if tensor is state_dict["module_pipeline.m_stage1.linear.weight"]:
        return flow.sbp.split(1)
    return flow.sbp.split(0)

state_dict = flow.utils.to_global(
    local_state_dict,
    placement=model_file_placement,
    sbp=get_sbp,
)
```

daquexian commented on python/oneflow/utils/global_view/to_global.py:
```diff
@@ -0,0 +1,179 @@
+    Args:
+        input (oneflow.Tensor/None/list/tuple/dict): the input that needs to be converted.
+        placement (oneflow.placement, optional): the desired placement of the input. Default: None
+        sbp (oneflow.sbp.sbp or list/tuple of oneflow.sbp.sbp, optional): the desired sbp of the input or self-defined functions in order to specify SBP. Default: None
```
这里 sbp 的类型可以添加函数类型。不知道在文档里怎么表示，也许是和 python type hinting 一致的 Callable[[Tensor], oneflow.sbp.sbp]

daquexian commented on python/oneflow/test/modules/test_global_cast.py:
```diff
@@ -13,16 +13,18 @@
+from oneflow import nn
 import oneflow.unittest
 from oneflow.test_utils.test_util import GenArgList
+from oneflow.framework.tensor import Tensor
```
这里可以直接 from oneflow import Tensor 吗

daquexian commented on python/oneflow/utils/global_view/to_global.py:
```diff
@@ -0,0 +1,179 @@
+    is_tensor_transform_list = 0
+    if isinstance(input, Tensor):
+        input = [input]
+        is_tensor_transform_list = 1
```
```suggestion
        is_tensor_transform_list = True
```

未来可以让 ArgsTree 直接支持 Tensor 吗，现在的处理方式感觉不太优雅

strint commented on python/oneflow/utils/__init__.py:
```diff
@@ -14,4 +14,6 @@
-from oneflow.utils.torch.from_or_to_torch_tensor import from_torch, to_torch
+from oneflow.utils.tensor.from_or_to_torch_tensor import from_torch, to_torch
+from oneflow.utils.global_view.to_global import to_global
+from oneflow.utils.global_view.to_local import to_local
```
删掉具体的接口，只 import global_view/tensor

strint commented on python/oneflow/utils/global_view/global_utils.py:
```diff
@@ -0,0 +1,80 @@
+        )
+    else:
+        if "grad_sbp" in kwargs:
+            del kwargs["grad_sbp"]
```
这里注释一下删掉的原因

lixiang007666 replied:
这个地方是参考：https://github.com/Oneflow-Inc/oneflow/blob/2717a4e147bd81d7ff7f36863d2f2456ceb8d95c/python/oneflow/nn/modules/global_cast.py#L83

strint commented on python/oneflow/utils/global_view/to_global.py:
```diff
@@ -0,0 +1,179 @@
+            if tensor is state_dict["module_pipeline"]["m_stage3.linear.weight"]:
+                return flow.sbp.split(1)
+            if tensor is state_dict["module_pipeline"]["m_stage3.linear.bias"]:
+                return BROADCAST
```
这里改成用 oneflow.sbp.broadcast

lixiang007666 commented on python/oneflow/utils/global_view/to_global.py:
```diff
@@ -0,0 +1,192 @@
+                    # For the use of the get_sbp function, broadcast the SBP of src_rank to other ranks, avoiding the situation of none is none.
+                    src_rank_sbp = src_sbp_broadcast(
+                        sbp(input, node), placement.ranks.flat[0]
+                    )
```
处理了今天上午提到的问题：比如在 rank 2,3（非 placement.ranks） 上广播后，input 是 value 为 None 的字典。
这时在后面的 map_leaf 过程中调用用户自定义的 get_sbp 函数会出现 `if none is none` 的情况，导致 SBP 总是 Broadcast，与 src_rank 的 SBP 不一致导致加载模型卡住。
修复了这种情况。

strint replied:
这里对每个 tensor，都做了一下 sbp 的广播。这个可以解决问题。不过坏处是多做了很多通信。

strint commented on python/oneflow/utils/global_view/to_global.py:
```diff
@@ -0,0 +1,192 @@
+
+                def leaf_fn_to_none(node):
+                    if isinstance(node, Tensor):
+                        return None
```
可以把这里替换成一个 flow.empty(0, 1) 的 0-size tensor，比较小。

这样每个 rank 都有 tensor 实例，可以避免 None 的问题。
